README.md


<div id="badges"><!-- pkgdown markup -->
<a href="https://CRAN.R-project.org/web/checks/check_results_R.matlab.html"><img border="0" src="https://www.r-pkg.org/badges/version/R.matlab" alt="CRAN check status"/></a> <a href="https://github.com/HenrikBengtsson/R.matlab/actions?query=workflow%3AR-CMD-check"><img border="0" src="https://github.com/HenrikBengtsson/R.matlab/actions/workflows/R-CMD-check.yaml/badge.svg?branch=develop" alt="R CMD check status"/></a>     <a href="https://app.codecov.io/gh/HenrikBengtsson/R.matlab"><img border="0" src="https://codecov.io/gh/HenrikBengtsson/R.matlab/branch/develop/graph/badge.svg" alt="Coverage Status"/></a> 
</div>

# R.matlab: Read and Write MAT Files and Call MATLAB from Within R 


## Installation
R package R.matlab is available on [CRAN](https://cran.r-project.org/package=R.matlab) and can be installed in R as:
```r
install.packages("R.matlab")
```


### Pre-release version

To install the pre-release version that is available in Git branch `develop` on GitHub, use:
```r
remotes::install_github("HenrikBengtsson/R.matlab", ref="develop")
```
This will install the package from source.  

<!-- pkgdown-drop-below -->


## Contributing

To contribute to this package, please see [CONTRIBUTING.md](CONTRIBUTING.md).



NEWS
Package: R.matlab
=================

Version: 3.6.2-9002 [2022-07-23]

DOCUMENTATION:

 * Updated URLs that were redirected to a new location.
 

Version: 3.6.2 [2018-09-26]

BUG FIXES:

 * Matlab$getOption() produced "Error in base::getOption(...) : 'x' must be a
   character string" in R-devel (to become R 3.6.0).  Thank you Kurt Hornik
   for the troubleshooting and the fix.
 
CODE REFACTORING:

 * CLEANUP: Drop code conditional on non-supported old versions of R.

 
Version: 3.6.1 [2016-10-19]

NEW FEATURES:

 * Now readMat() detects when MAT v7.3 files are trying to be read and gives an
   informative error explaining those are not supported (Issue #23).  Thanks to
   Thomas Beutlich (author of C library matio) for (non-official) pointers on
   how to distinguish MAT v5 from MAT v7.3 files.
   
 
Version: 3.6.0 [2016-07-05]

NEW FEATURES:

 * writeMat() gained argument 'fixNames=TRUE' for automatically replacing
   periods (.) in variable and field names with underscores (_). Thanks to
   Eugenio Piasini at University College London for suggesting this feature.

 * The MatlabServer.m script now outputs verbose/debug messages to standard
   error (or more precisely MATLAB file #2) (Issue #27).

 * Internal writeCommand() now gives an informative error if an unknown command
   is used.  Previously it silently sent a "0L" command to MATLAB.

SOFTWARE QUALITITY:

 * Increased test coverage to 87% (was 73%).

BUG FIXES:

 * writeMat() could not write logical elements.  They are now converted to
   integers as MATLAB does not have a native logical / boolean data type.
   Previously, writeMat() silently ignored writing the data section of these
   logical elements resulting in a MAT file that could not be parsed by MATLAB
   (although readMat() could).  Thanks to Eugenio Piasini for reporting on this
   problem.
 
DEPRECATED AND DEFUNCT:

 * CLEANUP: Exporting only methods parts of the public API.

 
Version: 3.5.1 [2016-03-27]

DOCUMENTATION:

 * Dropping reference to deprecated utils::setInternet2() in help("Matlab").
 
 
Version: 3.5.0 [2016-03-05]

NEW FEATURES:

 * Now readMat() parses (effectively skips) also objects of array class
   mxOPAQUE_CLASS (=17).  The official 'MAT-File Format R2015b' (Sept 2015)
   documentation does not describe this type.  The parsing/skipping was done
   by reverse engineering and clever guesses.  Thanks to GitHub user 'yakzan'
   for reporting on this type of MAT files.
    
SOFTWARE QUALITITY:

 * Package test coverage is 82%.

CODE REFACTORING:

 * CLEANUP: Internal fallback method for reading compressed MAT files using
   Omegahat's Rcompression package is not needed since R 2.10.0 and therefore
   dropped.

 
Version: 3.4.0 [2016-02-03]

NEW FEATURES:

 * Now readMat() parses also functional objects of array type (class) 16.  Note
   that the official 'MAT-File Format R2015b' (Sept 2015) documentation does
   not describe this type.  I had to implement the parser based on best guess
   on how the MAT file format works in general, so it might not work in all
   cases.  Thanks to GitHub user 'anilatx' for reporting on this type of MAT
   files.

 * ROBUSTNESS: Now writeMat(pathname, ...) writes the MAT file atomically by
   first writing to a temporary file which is renamed only if successful.

 * ROBUSTNESS: Now writeMat() asserts that a maximum of 2^31-1 bytes are
   written per variable, otherwise an informative error is thrown.  This byte
   limit is set by the MAT file format definition.  Thanks to GitHub user
   'intelinsight' for reporting on this.

SOFTWARE QUALITITY:

 * Package test coverage is 81%.
 
 
Version: 3.3.0 [2015-09-22]

NEW FEATURES:

 * evaluate() gained argument 'capture' for capturing MATLAB output while
   evaluating MATLAB expressions.  Thanks to Rohan Shah (GitHub user
   rohan-shah) for contributing code for this.

 * ROBUSTNESS: Now the temporary filenames generated by MATLAB also includes
   the port number in addition to a semi-random sequence string.  This should
   lower the risk for two MATLAB server using the same temporary filenames.

BUG FIXES:

 * readMat() gave an error when reading empty sparse matrices.  Thanks to
   Gregory Jefferis at University of Cambridge for the bug report and bug fix.

CODE REFACTORING:

 * Explicit import of 'methods' and 'utils' functions.


 
Version: 3.2.0 [2015-02-24]

NEW FEATURES:

 * Now Matlab$startServer() always overwrites any existing 'MatlabServer.m' and
   'InputStreamByteWrapper.class' to make sure the most recent versions are
   always used.  Previously, existing files would be kept, requiring a manual
   removal in order to use more resent versions of these files.

BUG FIXES:

 * readMat(..., sparseMatrixClass='matrix') did not return the correct results
   in all cases.  Added more package tests.

 * readMat(..., sparseMatrixClass='SparseM') did not handle all-zero sparse
   matrices.  Added package test for this.

 * readMat(..., sparseMatrixClass='Matrix') would give "Error in 
   validObject(.Object) : invalid class "dgCMatrix" object: lengths of slots
   'i' and 'x' must match", whenever trying to read an sparse matrix with all
   zeros.  Added package test for this case. Thanks to Abhinav Jauhri at
   Carnegie Mellon University for reporting on this and donating a test MAT
   file.

 * Trying to retrieve a non-existing MATLAB variable using getVariable() would
   crash the R-to-MATLAB communication if remote=FALSE.  Similarly, Trying to
   evaluate() a MATLAB expression that is invalid or by other means fails would
   also break the R-to-MATLAB bridge.  The only solution is both cases was to
   restart R and MATLAB.  Thanks to GitHub user badbye (yalei) for the report.
 
SOFTWARE QUALITITY:

 * Package test coverage is 77%.

 
Version: 3.1.1 [2014-10-10]

CODE REFACTORING:


R/old/Verbose.R
###########################################################################/**
# @RdocClass Verbose
#
# @title "Class for detailed verbose output with various details"
#
# \description{
#  @classhierarchy
# }
#
# @synopsis
#
# \arguments{
#   \item{threshold}{A threshold that the \code{level} argument
#     of any write method has to be equal to or larger than in order to the 
#     message being written.
#     Thus, the lower the threshold is the more and more details will be
#     outputted.}
# }
#
# \details{
#  @get "title".
# }
#
# @author
# 
# @keyword programming
# @keyword internal
#*/###########################################################################
setConstructorS3("Verbose", function(threshold=0) {
  if (!is.numeric(threshold) || length(threshold) != 1)
    throw("Argument 'threshold' must be a scalar.");
  extend(Object(), "Verbose",
    threshold = threshold
  )
}, private=TRUE)



###########################################################################/**
# @RdocMethod "as.character"
#
# @title "Returns a character string version of this object"
#
# \description{
#   @get "title".
# }
# 
# @synopsis
#
# \arguments{
#  \item{...}{Not used.}
# }
#
# \value{
#   Returns a @character string.
# }
#
# @author
#
# \seealso{
#   @seeclass
# }
#
# @keyword programming
#*/###########################################################################
setMethodS3("as.character", "Verbose", function(this, ...) {
  s <- paste(class(this)[1], ": threshold=", this$threshold, sep="");
  s;
})


###########################################################################/**
# @RdocMethod setThreshold
#
# @title "Sets verbose threshold"
#
# \description{
#   @get "title". Output requests below this threshold will be ignored.
# }
#
# @synopsis
#
# \arguments{
#  \item{threshold}{A @numeric threshold.}
#  \item{...}{Not used.}
# }
#
# \value{
#   Returns old threshold.
# }
#
# @author
#
# \seealso{
#   @seemethod "getThreshold" and @seemethod "isVisible".
#   @seeclass
# }
#
# @keyword programming
#*/###########################################################################
setMethodS3("setThreshold", "Verbose", function(this, threshold, ...) {
  if (!is.numeric(threshold) || length(threshold) != 1)
    throw("Argument 'threshold' must be a scalar.");
  old <- this$threshold;
  this$threshold <- threshold;
  invisible(old);
})


###########################################################################/**
# @RdocMethod getThreshold
#
# @title "Gets current verbose threshold"
#
# \description{
#   @get "title".
# }
#
# @synopsis
#
# \arguments{
#  \item{...}{Not used.}
# }
#
# \value{
#   Returns a @numeric value.
# }
#
# @author
#
# \seealso{
#   @seemethod "setThreshold" and @seemethod "isVisible".
#   @seeclass
# }
#
# @keyword programming
#*/###########################################################################
setMethodS3("getThreshold", "Verbose", function(this, ...) {
  this$threshold;
})


###########################################################################/**
# @RdocMethod isVisible
#
# @title "Checks if a certain verbose level will be shown or not"
#
# \description{
#   @get "title".
# }
# 
# @synopsis
#
# \arguments{
#  \item{level}{A @numeric value to be compared to the threshold.}
#  \item{...}{Not used.}
# }
#
# \value{
#   Returns @FALSE, if given level is below current threshold, otherwise
#   @FALSE.
# }
#
# @author
#
# \seealso{
#   @seemethod "getThreshold" and @seemethod "setThreshold".
#   @seeclass
# }
#
# @keyword programming
#*/###########################################################################
setMethodS3("isVisible", "Verbose", function(this, level, ...) {
  (level >= this$threshold);
})



###########################################################################/**
# @RdocMethod cat
#
# @title "Concatenates and prints objects if above threshold"
#
# \description{
#   @get "title".
# }
# 
# @synopsis
#
# \arguments{
#  \item{...}{Objects to be passed to @see "base::cat".}
#  \item{level}{A @numeric value to be compared to the threshold.}
# }
#
# \value{
#   Returns nothing.
# }
#
# @author
#
# \seealso{
#   @seeclass
# }
#
# @keyword programming
#*/###########################################################################
setMethodS3("cat", "Verbose", function(this, ..., level=0) {
  if (is.null(level) || isVisible(this, level)) {
    cat(...);
  }
})


###########################################################################/**
# @RdocMethod print
#
# @title "Prints objects if above threshold"
#
# \description{
#   @get "title".
# }
# 
# @synopsis
#
# \arguments{
#  \item{...}{Objects to be passed to @see "base::print".}
#  \item{level}{A @numeric value to be compared to the threshold.}
# }
#
# \value{
#   Returns nothing.
# }
#
# @author
#
# \seealso{
#   @seeclass
# }
#
# @keyword programming
#*/###########################################################################
setMethodS3("print", "Verbose", function(x, ..., level=0) {
  # print() is already a generic function in 'base'.
  this <- x;

  if (is.null(level) || isVisible(this, level)) {
    print(...);
  }
})


###########################################################################/**
# @RdocMethod str
#
# @title "Prints the structure of an object if above threshold"
#
# \description{
#   @get "title".
# }
# 
# @synopsis
#
# \arguments{
#  \item{...}{Objects to be passed to @see "base::str".}
#  \item{level}{A @numeric value to be compared to the threshold.}
# }
#
# \value{
#   Returns nothing.
# }
#
# @author
#
# \seealso{
#   @seeclass
# }
#
# @keyword programming
#*/###########################################################################
setMethodS3("str", "Verbose", function(object, ..., level=0) {
  # print() is already a generic function in 'base'.
  this <- object;

  if (is.null(level) || isVisible(this, level)) {
    str(...);
  }
})


###########################################################################/**
# @RdocMethod summary
#
# @title "Generates a summary of an object if above threshold"
#
# \description{
#   @get "title".
# }
# 
# @synopsis
#
# \arguments{
#  \item{...}{Objects to be passed to @see "base::summary".}
#  \item{level}{A @numeric value to be compared to the threshold.}
# }
#
# \value{
#   Returns nothing.
# }
#
# @author
#
# \seealso{
#   @seeclass
# }
#
# @keyword programming
#*/###########################################################################
setMethodS3("summary", "Verbose", function(object, ..., level=0) {
  # print() is already a generic function in 'base'.
  this <- object;

  if (is.null(level) || isVisible(this, level)) {
    print(summary(...));
  }
})

###########################################################################/**
# @RdocMethod printf
#
# @title "Formats and prints object if above threshold"
#
# \description{
#   @get "title".
# }
# 
# @synopsis
#
# \arguments{
#  \item{...}{Objects to be passed to @see "base::sprintf".}
#  \item{level}{A @numeric value to be compared to the threshold.}
# }
#
# \value{
#   Returns nothing.
# }
#
# @author
#
# \seealso{
#   @seeclass
# }
#*/###########################################################################
setMethodS3("printf", "Verbose", function(this, fmtstr, ..., level=0) {
  if (is.null(level) || isVisible(this, level)) {
    cat(sprintf(fmtstr, ...));
  }
})

###########################################################################/**
# @RdocMethod newline
#
# @title "Prints a newline if above threshold"
#
# \description{
#   @get "title".
# }
# 
# @synopsis
#
# \arguments{
#  \item{level}{A @numeric value to be compared to the threshold.}
#  \item{...}{Not used.}
# }
#
# \value{
#   Returns nothing.
# }
#
# @author
#
# \seealso{
#   @seeclass
# }
#*/###########################################################################
setMethodS3("newline", "Verbose", function(this, level=0, ...) {
  cat(this, "\n", level=level);
})


###########################################################################/**
# @RdocMethod ruler
#
# @title "Prints a ruler if above threshold"
#
# \description{
#   @get "title".
# }
# 
# @synopsis
#
# \arguments{
#  \item{chars}{A @character string concatenated to a ruler.}
#  \item{width}{An @numeric specifying the width (in characters) of 
#    the ruler.}
#  \item{level}{A @numeric value to be compared to the threshold.}
#  \item{...}{Not used.}
# }
#
# \value{
#   Returns nothing.
# }
#
# @author
#
# \seealso{
#   @seeclass
# }
#*/###########################################################################
setMethodS3("ruler", "Verbose", function(this, chars="- ", width=getOption("width")-1, level=0, ...) {
  if (is.null(level) || isVisible(this, level)) {
    chars <- strsplit(chars, "")[[1]];
    ruler <- rep(chars, length.out=width);
    ruler <- paste(ruler, collapse="");
    cat(ruler, "\n", sep="");
  }
})


###########################################################################/**
# @RdocMethod evaluate
#
# @title "Evaluates a function and prints its results if above threshold"
#
# \description{
#   @get "title".
# }
# 
# @synopsis
#
# \arguments{
#  \item{fun}{A @function to be evaluated (only if above threshold).}
#  \item{...}{Additional arguments passed to the function.}
#  \item{level}{A @numeric value to be compared to the threshold.}
# }
#
# \value{
#   Returns nothing.
# }
#
# @author
#
# \seealso{
#   @seeclass
# }
#*/###########################################################################
setMethodS3("evaluate", "Verbose", function(this, fun, ..., level=0) {
  if (is.null(level) || isVisible(this, level)) {
    cat(fun(...));
  }
})



############################################################################
# HISTORY:
# 2005-05-26
# o BUG FIX: Used 'threshold' instead of 'this$threshold' in as.character().
# 2005-02-15
# o Added arguments '...' in order to match any generic functions. 
# 2005-02-10
# o Added Rdoc comments to all of Verbose methods.
# 2004-02-24
# o Removed the requirement that the threshold must be non-negative.
# o The default verbose level is now 'level=0'.
# o Moved this class into the R.matlab package. Still a trial class that
#   may not exists later.
# 2003-12-07
# o Created.
############################################################################
